src/shared/watchdog.c

```diff
@@ -19,27 +19,30 @@ static char *watchdog_device = NULL;
 static usec_t watchdog_timeout = USEC_INFINITY;
 static usec_t watchdog_last_ping = USEC_INFINITY;
 
+static unsigned int usec_to_sec(usec_t val) {
+        usec_t t = DIV_ROUND_UP(val, USEC_PER_SEC);
+        return (unsigned int) t >= (UINT_MAX / 1000) ? (UINT_MAX / 1000) : t; /* Saturate to watchdog max */
+}
+
 static int update_timeout(void) {
         if (watchdog_fd < 0)
                 return 0;
         if (watchdog_timeout == USEC_INFINITY)
                 return 0;
 
         if (watchdog_timeout == 0) {
-                int flags;
+                unsigned int flags;
 
                 flags = WDIOS_DISABLECARD;
                 if (ioctl(watchdog_fd, WDIOC_SETOPTIONS, &flags) < 0)
                         return log_warning_errno(errno, "Failed to disable hardware watchdog: %m");
         } else {
                 char buf[FORMAT_TIMESPAN_MAX];
-                int sec, flags;
-                usec_t t;
+                unsigned int sec, flags;
 
-                t = DIV_ROUND_UP(watchdog_timeout, USEC_PER_SEC);
-                sec = (int) t >= INT_MAX ? INT_MAX : t; /* Saturate */
+                sec = usec_to_sec(watchdog_timeout);
                 if (ioctl(watchdog_fd, WDIOC_SETTIMEOUT, &sec) < 0)
```

Comment:
hmm, reading through the kernel sources, it actually uses "int" as type in copy_from_user() for most (all?) implementations of this ioctl. so this change seems to allow us passing negative values to the ioctl we previously didn't allow?

Reply:
similar for WDIOC_SETOPTIONS, the drivers I looked at all use "int" as type, but you change it to "unsigned" here?

Reply:
I see that the watchdog driver (`watchdog_ioctl`) uses `get_user` to copy the argument into an `unsigned int` for all of the watchdog ioctls (see: https://github.com/torvalds/linux/blame/50be9417e23af5a8ac860d998e1e3f06b8fd79d7/drivers/watchdog/watchdog_dev.c#L725). 

I think it makes sense to change it for `WDIOC_SETTIMEOUT` since it reflects how the kernel actually uses the variable. Regardless, `UINT_MAX / 1000` (4294967s) is smaller than `INT_MAX` (2147483647s) so we do not risk an integer overflow even if it was used as a signed value. But it also seems like if we ever did saturate the timeout to INT_MAX, the kernel would have rejected it anyway (at least on newer kernels).

For `WDIOC_SETOPTIONS`, we are performing bitwise operations exclusively so I think it doesn't make any functional difference but an unsigned integer better represents the actual usage of the variable (in my opinion).

Reply:
maybe maybe this limits explicitly by defining the limit as:

```c
#define WATCHDOG_TIMEOUT_MAX_SEC CONST_MIN(UINT_MAX/1000U, INT_MAX)
```

And then add a comment that drivers internally use "int", and the watchdog core enforces UINT_MAX/1000 as limit. Of course, we know only the UINT_MAX/1000 thing matters, but it's nice to encode in code what the though process is.

Reply:
I disagree since the timeout is really stored in the kernel as `unsigned int timeout;` (see [watchdog.h](https://elixir.bootlin.com/linux/latest/source/include/linux/watchdog.h#L102)). And the `watchdog_set_timeout` (the function that sets the timeout when we run the `WDIOC_SETTIMEOUT` ioctl) uses `unsigned int` as well (see [watchdog_dev.c](https://elixir.bootlin.com/linux/latest/source/drivers/watchdog/watchdog_dev.c#L390)):

```c
static int watchdog_set_timeout(struct watchdog_device *wdd, unsigned int timeout)
```

Additionally, the hardware driver interfaces use `unsigned int` as well (see [watchdog_ops](https://elixir.bootlin.com/linux/latest/source/include/linux/watchdog.h#L51)):

```c
int (*set_timeout)(struct watchdog_device *, unsigned int);
```

I do see that some drivers that do not use the new watchdog API do use signed integers for handling the timeouts (e.g. the [f71808e wdt](https://elixir.bootlin.com/linux/latest/source/drivers/watchdog/f71808e_wdt.c#L221) and the [mpcore wdt](https://elixir.bootlin.com/linux/v3.10.108/source/drivers/watchdog/mpcore_wdt.c#L139)). However, the watchdog maintainers are not accepting new drivers that do not use the new API and older drivers are actively [being updated](https://www.spinics.net/lists/linux-watchdog/msg20822.html) to use the new API. Since we know that UINT_MAX/1000 is smaller than INT_MAX anyway and the kernel is moving in this direction of using the new watchdog API, I think it makes sense to follow and assume unsigned integers are used under the hood.

Reply:
So yes, internally the kernel uses unsigned int, but the ioctl ifaces still are defined for "int", no? even the kernel docs say int, see Documentation/watchdog/watchdog-api.rst

I mean, I like expressing this situation with a CONST_MIN() expression, since that way we write in code that conceptually there are two limits applied (one internally in the kernel, one in the userspace/kernel api iface), and that the lower of the two of course matters.

I'd be fine to document this in a comment too (comments are always great!) but i like to express this in formalized C too to the level this makes sense, because it is concise and not slower (since the compiler reduces constant expressions anyway).